src/edu/tutorials/trainreservation/domain/Train.java

```diff
@@ -28,16 +26,32 @@ public List<Coach> getCoaches() {
     }
 
     public boolean hasRoute(String sourceCity, String destinationCity) {
-        return this.source.getName().equals(sourceCity)
-                && this.destination.getName().equals(destinationCity);
+        int sourceIndex = -1;
+        int destinationIndex = -1;
+
+        int i = 0;
+        while(sourceIndex == -1 || destinationIndex == -1) {
+            if (cities.get(i).getName().equals(sourceCity)) {
+                sourceIndex = i;
+            }
+            else if (cities.get(i).getName().equals(destinationCity)) {
+                destinationIndex = i;
+            }
+            if(i == cities.size() - 1) {
+                break;
+            }
+            i++;
+        }
+
+        return sourceIndex != -1 && destinationIndex != -1 && sourceIndex < destinationIndex;
```

Comment:
This code is functionally correct and includes edge cases, but we can improve its readability. Currently, reader needs to look into or debug the code to understand how it works and ensure it covers all scenarios.